engine.lisp

```diff
@@ -61,26 +61,44 @@ for the selector to the matcher."))
   ((%value :initarg :value :initform NIL :accessor value))
   (:documentation "Condition signalled to immediately return from MATCH-PAIR."))
 
-(defun split (split string)
-  (declare (optimize speed))
-  (macrolet ((body ()
-               `(loop with i = 0 and length = (length string)
-                      while (< i length)
-                      for item = (with-output-to-string (output)
-                                   (loop while (< i length)
-                                         for char = (aref string i)
-                                         do (cond ((char= char split)
-                                                   (loop while (and (< i length)
-                                                                    (char= (aref string i) split))
-                                                         do (incf i))
-                                                   (return))
-                                                  (T (write-char char output)
-                                                     (incf i)))))
-                      unless (string= item "")
-                      collect item)))
-    (typecase string
-      (simple-string (body))
-      (string (body)))))
+(defun split-member (item split string)
+  (declare (optimize speed)
+           (type string item string)
+           (type character split))
+  (macrolet ((with-stringcase (var body)
+               `(typecase ,var
+                  (simple-string ,body)
+                  (string ,body)))
+             (body ()
```

Comment:
`with-strincase` is already a macro that duplicates the body code for each case. I don't see how making the body another macro would help further.